vault/programs/vault/src/lib.rs

```diff
@@ -0,0 +1,464 @@
+        let seeds = vault_payer_seeds!(
+            vault_name = vault_name,
+            bump = ctx.accounts.vault.bumps.vault_payer
+        );
+        zeta_client::deposit(
+            ctx.accounts.zeta_program.clone(),
+            ctx.accounts.deposit_cpi_accounts.clone(),
+            seeds,
+            amount,
+        )
+    }
+
+    // #[access_control(deposit_withdraw_phase(&ctx.accounts.vault))]
+    pub fn withdraw_zeta(ctx: Context<WithdrawZeta>, amount: u64) -> ProgramResult {
+        let vault_name = ctx.accounts.vault.vault_name.as_ref();
+        let seeds = vault_payer_seeds!(
+            vault_name = vault_name,
+            bump = ctx.accounts.vault.bumps.vault_payer
+        );
+        zeta_client::withdraw(
+            ctx.accounts.zeta_program.clone(),
+            ctx.accounts.withdraw_cpi_accounts.clone(),
+            seeds,
+            amount,
+        )
+    }
+
+    pub fn initialize_zeta_open_orders(ctx: Context<InitializeZetaOpenOrders>) -> ProgramResult {
+        let vault_name = ctx.accounts.vault.vault_name.as_ref();
+        let seeds = vault_payer_seeds!(
+            vault_name = vault_name,
+            bump = ctx.accounts.vault.bumps.vault_payer
+        );
+        zeta_client::initialize_open_orders(
+            ctx.accounts.zeta_program.clone(),
+            ctx.accounts.initialize_open_orders_cpi_accounts.clone(),
+            seeds,
+        )
+    }
+
+    // TODO: in future move this on-chain
+    pub fn place_auction_order(
+        ctx: Context<PlaceAuctionOrder>,
+        price: u64,
+        size: u64,
+        side: Side,
+        client_order_id: Option<u64>,
+    ) -> ProgramResult {
+        msg!("PLACE AUCTION ORDER");
+        let vault_name = ctx.accounts.vault.vault_name.as_ref();
+        let seeds = vault_payer_seeds!(
+            vault_name = vault_name,
+            bump = ctx.accounts.vault.bumps.vault_payer
+        );
+        zeta_client::place_order(
+            ctx.accounts.zeta_program.clone(),
+            ctx.accounts.place_order_cpi_accounts.clone(),
+            seeds,
+            price,
+            size,
+            side,
+            client_order_id,
+        )
+    }
+
+    pub fn cancel_auction_order(
+        ctx: Context<CancelAuctionOrder>,
+        side: Side,
+        order_id: u128,
+    ) -> ProgramResult {
+        msg!("CANCEL AUCTION ORDER");
+        let vault_name = ctx.accounts.vault.vault_name.as_ref();
+        let seeds = vault_payer_seeds!(
+            vault_name = vault_name,
+            bump = ctx.accounts.vault.bumps.vault_payer
+        );
+        zeta_client::cancel_order(
+            ctx.accounts.zeta_program.clone(),
+            ctx.accounts.cancel_order_cpi_accounts.clone(),
+            seeds,
+            side,
+            order_id,
+        )
+    }
+
+    #[access_control(epoch_over(&ctx.accounts.vault))]
+    pub fn rollover_vault(
+        ctx: Context<RolloverVault>,
+        _vault_name: String,
```

Comment:
These args are already stored in the vault account

Reply:
Agree, but also how do I validate the seeds in the instruction? Surely can't read off vault_name from the vault for its own seed checks.
See `context.rs`:
```
    #[account(
        mut,
        seeds = [vault_name.as_bytes()],
        bump = bumps.vault,
        has_one = vault_admin
    )]
    pub vault: Box<Account<'info, Vault>>,
    ```

Reply:
That's the point though? Since the PDA you passed in matches it's expected contents. One can argue you don't need to check the seeds of a Vault since it's already been created.

Reply:
That's a very valid point, I'll cop that